wingtips-old-servlet-api/src/main/java/com/nike/wingtips/servlet/RequestTracingFilterServlet2Api.java

```diff
@@ -0,0 +1,247 @@
+package com.nike.wingtips.servlet;
+
+import com.nike.wingtips.Span;
+import com.nike.wingtips.TraceHeaders;
+import com.nike.wingtips.Tracer;
+import com.nike.wingtips.util.TracingState;
+
+import org.slf4j.Logger;
+import org.slf4j.LoggerFactory;
+
+import java.io.IOException;
+import java.util.ArrayList;
+import java.util.Collections;
+import java.util.List;
+
+import javax.servlet.Filter;
+import javax.servlet.FilterChain;
+import javax.servlet.FilterConfig;
+import javax.servlet.ServletException;
+import javax.servlet.ServletRequest;
+import javax.servlet.ServletResponse;
+import javax.servlet.http.HttpServletRequest;
+import javax.servlet.http.HttpServletResponse;
+
+import static com.nike.wingtips.util.AsyncWingtipsHelperJava7.unlinkTracingFromCurrentThread;
+
```

Comment:
Is there base class potential here such that servlet 3 and servlet 2 can share code and thus you could make this class abstract?  Could you push the common bits in this filter down into core as an abstract base class and then inherit into wingtips-servlet-api and wingtips-old-servlet-api?  

Or, if having servlet dependencies in core is no bueno, maybe have the abstract base in wingtips-servlet-api and the subclasses users would bind to be in wingtips-old-servlet-api and wingtips-servlet-api?

Also, I'm not so sure about having this class's name ending in "-Api", it seems to me that one would use this class directly as a servlet filter.  So, maybe just RequestTracingFilterServlet2?

Reply:
The servlet 2 filter is the base class in this scenario (it just doesn't need to be abstract), and the servlet 3 version only has to define the async bits - i.e. it shares most of the code of the servlet 2 filter already. Servlet dependency in core is definitely no bueno, and I don't see how further abstracting things helps in this case anyway.

I agree the name is unfortunate, but `RequestTracingFilterServlet2` sounds like it's "version 2" of the servlet filter, which is not the intent.

Reply:
I decided to punt and renamed it `RequestTracingFilterOldServlet`, which itself is not a great name, but at least it's not misleading (I hope?).

Reply:
I also changed the servlet API dependencies to `compileOnly`, similar to `provided` scope in maven, to prevent some runtime version conflicts with whatever servlet container it gets deployed to. This could lead to `NoClassDefFoundError`s in certain environments where you try to use these classes but don't have a servlet API dependency on your classpath, but the only realistic case I can think of where this might happen would be sub-libraries that build on these libs. This should be a rare case, and docs have been added to the submodule readmes explaining the situation. 

Not ideal, but it's kind of a pick-your-poison situation. I decided to go with preventing weird dependency version frustrations in the common case at the expense of being required to pull in a seems-to-be-missing dependency in the rare case.